Source/DafnyRuntime/DafnyRuntimeRust/src/lib.rs

```diff
@@ -870,62 +870,96 @@ where
         }
     }
 
-    pub fn append_recursive_safe(
-        array: &mut Vec<T>,
-        cache_borrow: &Option<&Rc<Vec<T>>>,
-        left: &Rc<RefCell<Sequence<T>>>,
-        right: &Rc<RefCell<Sequence<T>>>,
-    ) {
-        if let Some(values) = cache_borrow.as_ref() {
-            for value in values.iter() {
-                array.push(value.clone());
-            }
-            return;
-        }
-        #[cfg(not(feature = "sync"))]
-        {
-            Sequence::<T>::append_recursive(array, &left.as_ref().borrow());
-            Sequence::<T>::append_recursive(array, &right.as_ref().borrow());
-        }
-        #[cfg(feature = "sync")]
-        {
-            let left_guard = left.as_ref().lock().unwrap();
-            let right_guard = right.as_ref().lock().unwrap();
-            Sequence::<T>::append_recursive(array, &left_guard);
-            Sequence::<T>::append_recursive(array, &right_guard);
-        }
-    }
-
-    pub fn append_recursive(array: &mut Vec<T>, this: &Sequence<T>) {
+    pub fn append_simple(array: &mut Vec<T>, this: &Sequence<T>) -> bool {
         match this {
             Sequence::ArraySequence { values, .. } =>
-            // The length of the elements
             {
                 for value in values.iter() {
                     array.push(value.clone());
                 }
+                true
             }
-            Sequence::ConcatSequence {
-                cache: boxed, left, right, ..
-            } =>
-            // Let's create an array of size length and fill it up recursively
-            {
-                #[cfg(feature = "sync")]
-                let into_boxed = boxed.as_ref();
-                #[cfg(feature = "sync")]
-                let into_boxed_borrowed = into_boxed;
-                #[cfg(feature = "sync")]
-                let guard = into_boxed_borrowed.lock().unwrap();
-                #[cfg(feature = "sync")]
-                let borrowed: Option<&Rc<Vec<T>>> = guard.as_ref();
+            _ => false
+        }
+    }
 
-                #[cfg(not(feature = "sync"))]
-                let into_boxed = boxed.as_ref().clone();
-                #[cfg(not(feature = "sync"))]
-                let into_boxed_borrowed = into_boxed.borrow();
-                #[cfg(not(feature = "sync"))]
-                let borrowed: Option<&Rc<Vec<T>>> = into_boxed_borrowed.as_ref();
-                Self::append_recursive_safe(array, &borrowed, left, right);
+    pub fn append_recursive(array: &mut Vec<T>, this: &Sequence<T>) {
+            let mut this_ptr : * const Sequence<T> = this;
+        loop {
+            match unsafe { &*this_ptr  } {
```

Comment:
Please insert a `//SAFETY: ...` comment before this unsafe even if it is obvious, e.g.:
```suggestion
            // SAFETY: The invariant is that this_ptr is always a pointer to a sub-sequence of the original `this`parameter when following the field name 'right', which are allocated in this scope
            match unsafe { &*this_ptr  } {
```